File: C++/source/Framework/ObjectModel/WaveElement.cpp
/***************************************************************************
 *   Copyright (C) 2005-2013 Vidyasagara Guntaka                           *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "Framework/ObjectModel/WaveElement.h"
#include "Framework/Messaging/Local/WaveMessage.h"
#include "Framework/Utils/WaveAsynchronousContext.h"
#include "Framework/ObjectModel/WaveObjectManager.h"
#include "Framework/ObjectModel/WaveObjectManagerCommitTransactionContext.h"
#include "Framework/Utils/TraceUtils.h"
#include "Framework/Utils/AssertUtils.h"

#include "Framework/Messaging/LightHouse/LightPulse.h"
namespace WaveNs
{

WaveElement::WaveElement (WaveObjectManager *pWaveObjectManager)
    : m_pWaveObjectManager (pWaveObjectManager)
{
    if (NULL == m_pWaveObjectManager)
    {
        // THIS MUST NEVER HAPPEN
        // We cannot use trace or waveAssert here since they are abstract virtual here
        cerr << "Waveelement::Waveelement : You cannot create a WaveElement with NULL WaveObjectManager." << endl;
        assert (0);
    }
}

WaveElement::WaveElement (const WaveElement &waveElement)
{
// FIXME : sagar : Once we the cluster service fixes the NodeManagedObject copying by value into its vectors, enable the fatal statements below.

#if 0
    WaveNs::trace (TRACE_LEVEL_FATAL, "WaveElement::WaveElement : Copy Constructing a Wave Element does not make sense and hence not allowed.");
    WaveNs::waveAssert (false, __FILE__, __LINE__);
#else
    m_pWaveObjectManager = waveElement.m_pWaveObjectManager;
#endif
}

WaveElement::~WaveElement ()
{
}

WaveElement &WaveElement::operator = (const WaveElement &waveElement)
{
    WaveNs::trace (TRACE_LEVEL_FATAL, "WaveElement::operator = : Assigning to a Wave Element does not make sense and hence not allowed.");
    WaveNs::waveAssert (false, __FILE__, __LINE__);

    return (*this);
}

/**
 * Note, if any transaction was started before calling this step, and not committed,
 * the transaction will be commited to preserve the semantics, even though if there is
 * no db update by the step and then the reply will happen as usual.
 */
void WaveElement::waveLinearSequencerSucceededStep (WaveLinearSequencerContext *pWaveLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerSucceededStep : Entering ...");

    WaveMessage             *pWaveMessage             = pWaveLinearSequencerContext->getPWaveMessage ();
    WaveAsynchronousContext *pWaveAsynchronousContext = pWaveLinearSequencerContext->getPWaveAsynchronousContext ();

    if (true == (pWaveLinearSequencerContext->getIsHoldAllRequested ()))
    {
        pWaveLinearSequencerContext->unholdAll ();
    }

    ResourceId status = FRAMEWORK_SUCCESS;

    if (true == (pWaveLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveLinearSequencerContext->setIsTransactionStartedByMe (false);

        // commit the transaction to preserve the semantics of the executeSuccessStep and if the transaction
        // has no data to be committed, framework will immediately return without having to go to DB.
        status = commitTransaction();
    }

    if (FRAMEWORK_SUCCESS == status)
    {
       status = WAVE_MESSAGE_SUCCESS;
    }
    else
    {
        status = WAVE_COMMIT_TRANSACTION_FAILED;
    }

    if (NULL != pWaveMessage)
    {
        pWaveMessage->setCompletionStatus (status);
        reply (pWaveMessage);
    }
    else if (NULL != pWaveAsynchronousContext)
    {
        pWaveAsynchronousContext->setCompletionStatus (status);
        pWaveAsynchronousContext->callback ();
    }

    delete pWaveLinearSequencerContext;
}

void WaveElement::waveLinearSequencerFailedStep (WaveLinearSequencerContext *pWaveLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerFailedStep : Entering ...");

    WaveMessage             *pWaveMessage             = pWaveLinearSequencerContext->getPWaveMessage ();
    WaveAsynchronousContext *pWaveAsynchronousContext = pWaveLinearSequencerContext->getPWaveAsynchronousContext ();

    if (true == (pWaveLinearSequencerContext->getIsHoldAllRequested ()))
    {
        pWaveLinearSequencerContext->unholdAll ();
    }

    if (true == (pWaveLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveLinearSequencerContext->setIsTransactionStartedByMe (false);
        rollbackTransaction ();
    }

    if (NULL != pWaveMessage)
    {
        pWaveMessage->setCompletionStatus (pWaveLinearSequencerContext->getCompletionStatus ());
        reply (pWaveMessage);
    }
    else if (NULL != pWaveAsynchronousContext)
    {
        pWaveAsynchronousContext->setCompletionStatus (pWaveLinearSequencerContext->getCompletionStatus ());
        pWaveAsynchronousContext->callback ();
    }

    delete pWaveLinearSequencerContext;
}

void WaveElement::waveLinearSequencerStartTransactionStep (WaveLinearSequencerContext *pWaveLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerStartTransactionStep : Entering ...");

    if (false == (isTransactionInProgress ()))
    {
        pWaveLinearSequencerContext->setIsTransactionStartedByMe (true);

        startTransaction ();
    }

    pWaveLinearSequencerContext->executeNextStep (WAVE_MESSAGE_SUCCESS);
}

void WaveElement::waveLinearSequencerCommitTransactionStep (WaveLinearSequencerContext *pWaveLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerCommitTransactionStep : Entering ...");

    ResourceId status = WAVE_MESSAGE_SUCCESS;

    if (true == (pWaveLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveLinearSequencerContext->setIsTransactionStartedByMe (false);

        WaveObjectManagerCommitTransactionContext commitTransactionContext (pWaveLinearSequencerContext);

        status = commitTransaction (&commitTransactionContext);

        if (FRAMEWORK_SUCCESS == status)
        {
            status = WAVE_MESSAGE_SUCCESS;
        }
    }

    pWaveLinearSequencerContext->executeNextStep (status);
}

/**
 * Note, if any transaction was started before calling this step, and not committed,
 * the transaction will be committed to preserve the semantics, even though if there is
 * no db update by the step and then the reply will happen as usual.
 */
ResourceId WaveElement::waveSynchronousLinearSequencerSucceededStep (WaveSynchronousLinearSequencerContext *pWaveSynchronousLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveSynchronousLinearSequencerSucceededStep : Entering ...");

    WaveMessage             *pWaveMessage             = pWaveSynchronousLinearSequencerContext->getPWaveMessage ();
    WaveAsynchronousContext *pWaveAsynchronousContext = pWaveSynchronousLinearSequencerContext->getPWaveAsynchronousContext ();
    ResourceId                completionStatus          = pWaveSynchronousLinearSequencerContext->getCompletionStatus ();

    if (true == (pWaveSynchronousLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveSynchronousLinearSequencerContext->setIsTransactionStartedByMe (false);

        // commit the transaction to preserve the semantics of the WAVE_SEQUENCER_SKIP_TO_SUCCESS_STEP return code and if
        // the transaction has no data to be committed, framework will immediately return without having to go to DB.
        WaveObjectManagerCommitTransactionContext commitTransactionContext (pWaveSynchronousLinearSequencerContext);

        ResourceId status = commitTransaction (&commitTransactionContext);

        if (FRAMEWORK_SUCCESS != status)
        {
            completionStatus = WAVE_COMMIT_TRANSACTION_FAILED;
        }
    }

    if (NULL != pWaveMessage)
    {
        pWaveMessage->setCompletionStatus (completionStatus);
        reply (pWaveMessage);
    }
    else if (NULL != pWaveAsynchronousContext)
    {
        pWaveAsynchronousContext->setCompletionStatus (completionStatus);
        pWaveAsynchronousContext->callback ();
    }

    delete pWaveSynchronousLinearSequencerContext;
    return (completionStatus);
}

ResourceId WaveElement::waveSynchronousLinearSequencerFailedStep (WaveSynchronousLinearSequencerContext *pWaveSynchronousLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveSynchronousLinearSequencerFailedStep : Entering ...");

    WaveMessage             *pWaveMessage             = pWaveSynchronousLinearSequencerContext->getPWaveMessage ();
    WaveAsynchronousContext *pWaveAsynchronousContext = pWaveSynchronousLinearSequencerContext->getPWaveAsynchronousContext ();
    ResourceId                completionStatus          = pWaveSynchronousLinearSequencerContext->getCompletionStatus ();

    if (true == (pWaveSynchronousLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveSynchronousLinearSequencerContext->setIsTransactionStartedByMe (false);
        rollbackTransaction ();
    }

    if (NULL != pWaveMessage)
    {
        pWaveMessage->setCompletionStatus (completionStatus);
        reply (pWaveMessage);
    }
    else if (NULL != pWaveAsynchronousContext)
    {
        pWaveAsynchronousContext->setCompletionStatus (completionStatus);
        pWaveAsynchronousContext->callback ();
    }

    delete pWaveSynchronousLinearSequencerContext;
    return (completionStatus);
}

ResourceId WaveElement::waveSynchronousLinearSequencerStartTransactionStep (WaveSynchronousLinearSequencerContext *pWaveSynchronousLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerStartTransactionStep : Entering ...");

    if (false == (isTransactionInProgress ()))
    {
        pWaveSynchronousLinearSequencerContext->setIsTransactionStartedByMe (true);

        startTransaction ();
    }

    return (WAVE_MESSAGE_SUCCESS);
}

ResourceId WaveElement::waveSynchronousLinearSequencerCommitTransactionStep (WaveSynchronousLinearSequencerContext *pWaveSynchronousLinearSequencerContext)
{
    trace (TRACE_LEVEL_DEVEL, "WaveElement::waveLinearSequencerCommitTransactionStep : Entering ...");

    ResourceId status = WAVE_MESSAGE_SUCCESS;

    if (true == (pWaveSynchronousLinearSequencerContext->getIsTransactionStartedByMe ()))
    {
        pWaveSynchronousLinearSequencerContext->setIsTransactionStartedByMe (false);

        WaveObjectManagerCommitTransactionContext commitTransactionContext (pWaveSynchronousLinearSequencerContext);

        status = commitTransaction (&commitTransactionContext);

        if (FRAMEWORK_SUCCESS == status)
        {
            status = WAVE_MESSAGE_SUCCESS;
        }
    }

    return (status);
}

WaveObjectManager *WaveElement::getPWaveObjectManager () const
{
    return (m_pWaveObjectManager);
}

WaveManagedObject *WaveElement::createManagedObjectInstance (const string &managedClassName)
{
    trace (TRACE_LEVEL_ERROR, "WaveElement::createManagedObjectInstance : NOT IMPLEMENTED.  RETURNS NULL BY DEFAULT.");
    trace (TRACE_LEVEL_ERROR, "WaveElement::createManagedObjectInstance : ObjectManagers and Workers MUST overwrite this virtual function.");
    return (NULL);
}

WaveMessage *WaveElement::createMessageInstance (const UI32 &operationCode)
{
    trace (TRACE_LEVEL_ERROR, "WaveElement::createMessageInstance : NOT IMPLEMENTED.  RETURNS NULL BY DEFAULT.");
    trace (TRACE_LEVEL_ERROR, "WaveElement::createMessageInstance : ObjectManagers and Workers MUST overwrite this virtual function.");
    return (NULL);
}

LightPulse *WaveElement::createLightPulseInstance (const string &lightPulseName)
{
    trace (TRACE_LEVEL_ERROR, "WaveElement::createLightPulseInstance : NOT IMPLEMENTED.  RETURNS NULL BY DEFAULT.");
    trace (TRACE_LEVEL_ERROR, "WaveElement::createLightPulseInstance : ObjectManagers and Workers MUST overwrite this virtual function.");
    return (NULL);
}

void WaveElement::setPWaveObjectManager (WaveObjectManager *pWaveObjectManager)
{
    m_pWaveObjectManager = pWaveObjectManager;
}

}

File: Java/source/com/CxWave/Wave/Framework/ObjectModel/WaveElement.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.Framework.ObjectModel;

import com.CxWave.Wave.Framework.Messaging.LightHouse.LightPulse;
import com.CxWave.Wave.Framework.Messaging.Local.WaveEvent;
import com.CxWave.Wave.Framework.Messaging.Local.WaveMessage;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.NonEventHandler;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.NonLightPulseHandler;
import com.CxWave.Wave.Framework.ObjectModel.Annotations.NonMessageHandler;
import com.CxWave.Wave.Framework.Type.LocationId;
import com.CxWave.Wave.Framework.Type.TimeValue;
import com.CxWave.Wave.Framework.Type.TimerHandle;
import com.CxWave.Wave.Framework.Type.UI32;
import com.CxWave.Wave.Framework.Utils.Context.WaveAsynchronousContext;
import com.CxWave.Wave.Framework.Utils.Sequencer.WaveLinearSequencerContext;
import com.CxWave.Wave.Framework.Utils.Sequencer.WaveSynchronousLinearSequencerContext;
import com.CxWave.Wave.Resources.ResourceEnums.ResourceId;
import com.CxWave.Wave.Resources.ResourceEnums.TraceLevel;
import com.CxWave.Wave.Resources.ResourceEnums.WaveMessageStatus;

public abstract class WaveElement
{
    protected WaveObjectManager m_waveObjectManager;

    protected WaveElement ()
    {
    }

    protected WaveElement (final WaveObjectManager waveObjectManager)
    {
        m_waveObjectManager = waveObjectManager;
    }

    protected WaveElement (final WaveElement waveElement)
    {
        m_waveObjectManager = waveElement.m_waveObjectManager;
    }

    protected WaveObjectManager getWaveObjectManager ()
    {
        return (m_waveObjectManager);
    }

    protected void setWaveObjectManager (final WaveObjectManager waveObjectManager)
    {
        m_waveObjectManager = waveObjectManager;
    }

    protected abstract void trace (final TraceLevel requestedTraceLevel, final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix);

    protected void unknownTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_UNKNOWN, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void develTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_DEVEL, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void debugTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_DEBUG, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void infoTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_INFO, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void successTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_SUCCESS, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void warnTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_WARN, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void errorTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_ERROR, stringToTrace, addNewLine, suppressPrefix);
    }

    protected void fatalTrace (final String stringToTrace, final boolean addNewLine, final boolean suppressPrefix)
    {
        trace (TraceLevel.TRACE_LEVEL_FATAL, stringToTrace, addNewLine, suppressPrefix);
    }

    protected abstract void trace (final TraceLevel requestedTraceLevel, final String stringToTrace);

    protected void unknownTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_UNKNOWN, stringToTrace);
    }

    protected void develTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_DEVEL, stringToTrace);
    }

    protected void debugTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_DEBUG, stringToTrace);
    }

    protected void infoTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_INFO, stringToTrace);
    }

    protected void successTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_SUCCESS, stringToTrace);
    }

    protected void warnTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_WARN, stringToTrace);
    }

    protected void errorTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_ERROR, stringToTrace);
    }

    protected void fatalTrace (final String stringToTrace)
    {
        trace (TraceLevel.TRACE_LEVEL_FATAL, stringToTrace);
    }

    protected abstract void tracePrintf (final TraceLevel requestedTraceLevel, final boolean addNewLine, final boolean suppressPrefix, final String formatString, Object... objects);

    protected void unknownTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_UNKNOWN, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void develTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_DEVEL, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void debugTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_DEBUG, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void infoTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_INFO, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void successTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_SUCCESS, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void warnTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_WARN, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void errorTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_ERROR, addNewLine, suppressPrefix, formatString, objects);
    }

    protected void fatalTracePrintf (final boolean addNewLine, final boolean suppressPrefix, final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_FATAL, addNewLine, suppressPrefix, formatString, objects);
    }

    protected abstract void tracePrintf (final TraceLevel requestedTraceLevel, final String formatString, Object... objects);

    protected void unknownTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_UNKNOWN, formatString, objects);
    }

    protected void develTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_DEVEL, formatString, objects);
    }

    protected void debugTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_DEBUG, formatString, objects);
    }

    protected void infoTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_INFO, formatString, objects);
    }

    protected void successTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_SUCCESS, formatString, objects);
    }

    protected void warnTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_WARN, formatString, objects);
    }

    protected void errorTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_ERROR, formatString, objects);
    }

    protected void fatalTracePrintf (final String formatString, final Object... objects)
    {
        tracePrintf (TraceLevel.TRACE_LEVEL_FATAL, formatString, objects);
    }

    protected abstract void holdMessages ();

    protected abstract void holdHighPriorityMessages ();

    protected abstract void holdEvents ();

    public abstract void holdAll ();

    protected abstract void unholdMessages ();

    protected abstract void unholdHighPriorityMessages ();

    protected abstract void unholdEvents ();

    public abstract void unholdAll ();

    public abstract void updateTimeConsumedInThisThread (final UI32 operationCode, final int currentStep, final long lastLapDuration);

    public abstract void updateRealTimeConsumedInThisThread (final UI32 operationCode, final int currentStep, final long lastLapDuration);

    /**
     * Note, if any transaction was started before calling this step, and not committed, the transaction will be committed to
     * preserve the semantics, even though if there is no db update by the step and then the reply will happen as usual.
     */

    protected void waveLinearSequencerSucceededStep (final WaveLinearSequencerContext waveLinearSequencerContext)
    {
        develTrace ("WaveElement.waveLinearSequencerSucceededStep : Entering ...");

        final WaveMessage waveMessage = waveLinearSequencerContext.getWaveMessage ();
        final WaveAsynchronousContext waveAsynchronousContext = waveLinearSequencerContext.getWaveAsynchronousContext ();
        ResourceId completionStatus = waveLinearSequencerContext.getCompletionStatus ();

        if (true == (waveLinearSequencerContext.getIsHoldAllRequested ()))
        {
            waveLinearSequencerContext.unholdAll ();
        }

        if (true == (waveLinearSequencerContext.getIsTransactionStartedByMe ()))
        {
            waveLinearSequencerContext.setIsTransactionStartedByMe (false);

            // commit the transaction to preserve the semantics of the executeSuccessStep and if the transaction
            // has no data to be committed, framework will immediately return without having to go to DB.

            completionStatus = commitTransaction ();

            if (ResourceId.FRAMEWORK_SUCCESS.equals (completionStatus))
            {
                completionStatus = ResourceId.WAVE_MESSAGE_SUCCESS;
            }
            else
            {
                completionStatus = ResourceId.WAVE_COMMIT_TRANSACTION_FAILED;
            }
        }

        if (null != waveMessage)
        {
            waveMessage.setCompletionStatus (completionStatus);
            reply (waveMessage);
        }
        else if (null != waveAsynchronousContext)
        {
            waveAsynchronousContext.setCompletionStatus (completionStatus);
            waveAsynchronousContext.callback ();
        }
    }

    protected void waveLinearSequencerFailedStep (final WaveLinearSequencerContext waveLinearSequencerContext)
    {
        develTrace ("WaveElement::waveLinearSequencerFailedStep : Entering ...");

        final WaveMessage waveMessage = waveLinearSequencerContext.getWaveMessage ();
        final WaveAsynchronousContext waveAsynchronousContext = waveLinearSequencerContext.getWaveAsynchronousContext ();

        if (true == (waveLinearSequencerContext.getIsHoldAllRequested ()))
        {
            waveLinearSequencerContext.unholdAll ();
        }

        if (true == (waveLinearSequencerContext.getIsTransactionStartedByMe ()))
        {
            waveLinearSequencerContext.setIsTransactionStartedByMe (false);
            rollbackTransaction ();
        }

        if (null != waveMessage)
        {
            waveMessage.setCompletionStatus (waveLinearSequencerContext.getCompletionStatus ());
            reply (waveMessage);
        }
        else if (null != waveAsynchronousContext)
        {
            waveAsynchronousContext.setCompletionStatus (waveLinearSequencerContext.getCompletionStatus ());
            waveAsynchronousContext.callback ();
        }
    }

    protected abstract void rollbackTransaction ();

    @NonMessageHandler
    protected abstract ResourceId reply (final WaveMessage waveMessage);

    protected abstract ResourceId commitTransaction ();

    public WaveMessage createMessageInstance (final UI32 operationCode)
    {
        if (null != m_waveObjectManager)
        {
            return (m_waveObjectManager.createMessageInstance (operationCode));
        }
        else
        {
            return (null);
        }
    }

    protected ResourceId waveSynchronousLinearSequencerSucceededStep (final WaveSynchronousLinearSequencerContext waveSynchronousLinearSequencerContext)
    {
        infoTracePrintf ("WaveElement.waveSynchronousLinearSequencerSucceededStep : Entering ...");

        final WaveMessage waveMessage = waveSynchronousLinearSequencerContext.getWaveMessage ();
        final WaveAsynchronousContext waveAsynchronousContext = waveSynchronousLinearSequencerContext.getWaveAsynchronousContext ();
        ResourceId completionStatus = waveSynchronousLinearSequencerContext.getCompletionStatus ();

        if (true == (waveSynchronousLinearSequencerContext.getIsTransactionStartedByMe ()))
        {
            waveSynchronousLinearSequencerContext.setIsTransactionStartedByMe (false);

            // commit the transaction to preserve the semantics of the executeSuccessStep and if the transaction
            // has no data to be committed, framework will immediately return without having to go to DB.

            completionStatus = commitTransaction ();

            if (ResourceId.FRAMEWORK_SUCCESS.equals (completionStatus))
            {
                completionStatus = ResourceId.WAVE_MESSAGE_SUCCESS;
            }
            else
            {
                completionStatus = ResourceId.WAVE_COMMIT_TRANSACTION_FAILED;
            }
        }

        if (null != waveMessage)
        {
            waveMessage.setCompletionStatus (completionStatus);
            reply (waveMessage);
        }
        else if (null != waveAsynchronousContext)
        {
            waveAsynchronousContext.setCompletionStatus (completionStatus);
            waveAsynchronousContext.callback ();
        }

        return (completionStatus);
    }

    protected ResourceId waveSynchronousLinearSequencerFailedStep (final WaveSynchronousLinearSequencerContext waveSynchronousLinearSequencerContext)
    {
        infoTracePrintf ("WaveElement::waveLinearSequencerFailedStep : Entering ...");

        final WaveMessage waveMessage = waveSynchronousLinearSequencerContext.getWaveMessage ();
        final WaveAsynchronousContext waveAsynchronousContext = waveSynchronousLinearSequencerContext.getWaveAsynchronousContext ();
        final ResourceId completionStatus = waveSynchronousLinearSequencerContext.getCompletionStatus ();

        if (true == (waveSynchronousLinearSequencerContext.getIsTransactionStartedByMe ()))
        {
            waveSynchronousLinearSequencerContext.setIsTransactionStartedByMe (false);
            rollbackTransaction ();
        }

        if (null != waveMessage)
        {
            waveMessage.setCompletionStatus (completionStatus);
            reply (waveMessage);
        }
        else if (null != waveAsynchronousContext)
        {
            waveAsynchronousContext.setCompletionStatus (completionStatus);
            waveAsynchronousContext.callback ();
        }

        return (completionStatus);
    }

    @NonMessageHandler
    protected abstract WaveMessageStatus sendOneWay (final WaveMessage waveMessage);

    protected abstract WaveMessageStatus sendOneWay (final WaveMessage waveMessage, final LocationId locationId);

    @NonMessageHandler
    protected abstract WaveMessageStatus sendSynchronously (final WaveMessage waveMessage);

    protected abstract WaveMessageStatus sendSynchronously (final WaveMessage waveMessage, final LocationId locationId);

    protected abstract WaveMessageStatus send (final WaveMessage waveMessage, final WaveMessageResponseHandler waveMessageCallback, final Object waveMessageContext);

    protected abstract WaveMessageStatus send (final WaveMessage waveMessage, final WaveMessageResponseHandler waveMessageCallback, final Object waveMessageContext, final long timeOutInMilliSeconds, final LocationId locationId, final WaveElement waveMessageSender);

    @NonEventHandler
    protected abstract WaveMessageStatus broadcast (final WaveEvent waveEvent);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final TimeValue startInterval, final TimeValue periodicInterval, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext, final WaveElement waveTimerSender);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final TimeValue startInterval, final TimeValue periodicInterval, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContexts);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final TimeValue startInterval, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext, final WaveElement waveTimerSender);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final TimeValue startInterval, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final long startIntervalMilliSeconds, final long periodicIntervalMilliSeconds, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext, final WaveElement waveTimerSender);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final long startIntervalMilliSeconds, final long periodicIntervalMilliSeconds, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final long startIntervalMilliSeconds, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext, final WaveElement waveTimerSender);

    protected abstract ResourceId startTimer (final TimerHandle timerHandle, final long startIntervalMilliSeconds, final WaveTimerExpirationHandler waveTimerExpirationCallback, final Object waveTimerExpirationContext);

    protected abstract ResourceId deleteTimer (final TimerHandle timerHandle);

    @NonLightPulseHandler
    protected abstract WaveMessageStatus broadcastLightPulse (final LightPulse lightPulse);
}
